### ISSUE TITLE
Connecting to PostgreSQL

### ISSUE THREAD

**Issue by nawazahmed1** [cavaliercoder/libzbxpgsql#73]
Hi, 

I am setting up Zabbix monitoring (3.0.7) on a CentOS (7.3.1611) and following the online documentation. I installed zabbix-server-pgsql-3.0.7-1.el7.x86_64.rpm, zabbix-web-3.0.7-1.el7.noarch.rpm and zabbix-web-pgsql-3.0.7-1.el7.noarch.rpm packages on a server (A) and installed zabbix-agent-3.0.7-1.el7.x86_64.rpm on server (B). 

The front end is loaded up and i can login and added server B on the configuration -> Hosts and it shows up on the "Dashboard". I at "Connecting to PostgreSQL" of online documentation and it says something about the "pg.connect[<connection_string>,<database>]" and "pg.setting[<connection_string>,<database>,<setting>]". I got no clue where is this configured or entered. Could you please send me a screenshot/steps to get this configured. 

I am new to Zabbix, so most of the stuff i come up on the online doco. is a bit hassling to figure out. I am following the doco. from cover to cover to go step by step to have foolproof monitoring setup. Please help.

Nawaz Ahmed

**Comment by robbrucks** [comment]
Hi @nawazahmed1,

After installing the Zabbix Agent on your DB server you'll need to follow the steps on this page:
[libzbxpgsql Installation](http://cavaliercoder.com/libzbxpgsql/documentation/installation/)

The remaining steps will be to link the template to your host in Zabbix, and configure the connection macro settings for the host in Zabbix if needed.

Most of the remaining documentation is on setting up your own custom metrics.

Happy monitoring!
Rob

**Comment by robbrucks** [comment]
To configure the host macro variables:

Login to the Zabbix UI as an administrator
Click on "Configuration" menu
Click on "Hosts" sub-menu
Click on the name of your server
Click on the "Macros" sub-tab
Click on "Inherited and host macros" box

You should see the {$PG_CONN} variable and be able to change it from the default if needed.  But the default should work if you configured the pg_hba.conf (and the listener is enabled in postgresql.conf) as described in the docs.

You should be able to add any valid PostgreSQL connection parameter from the [PostgreSQL libpq Docs](https://www.postgresql.org/docs/9.2/static/libpq-connect.html), down in the "Parameter Keywords" section.

One word of caution: don't change macro variables on the template itself, only on the hosts linked to the template.

--Rob

**Comment by nawazahmed1** [comment]
Hi Rob,

Thank you so much for the swift response and the steps to follow. I apologize, i missed the part of loading up libzbxpgsql template on zabbix in my query.  I did follow the "Template installation" step and everything was fine..

I followed your elaborated steps above and I could see the {$PG_CONN} variable under the "Macros" subtab of the configured host. Here is what i found

{$PG_CONN} => host=10.44.37.48 port=30000 user=fsep connect_timeout=10
{$PG_DB} => db1

The user "fsep" is the superuser/database owner of the postgresql installation. Like you had mentioned, everything had been configured by default.

I am at the "Monitoring role" section of the chapter "Connecting to PostgreSQL" and followed as it says by creating a monitoring role etc. The section suggests to connect as the "monitoring" account. In this case, would my $PG_CONN macro have a value as below ?

{$PG_CONN} => host=10.44.37.48 port=30000 user=**monitoring,fsep** connect_timeout=10

Also, a follow up question from your response. > One word of caution: don't change macro variables on the template itself, only on the hosts linked to the template.

If i change anything on the "Inherited and host macros" then would that cover the caution above ?

Thanks & Regards,
Nawaz

**Comment by robbrucks** [comment]
Hi Nawaz,

What monitoring userid are you trying to use?  Is it "monitoring" or "fsep"?  You can only specify one userid in the connection string, and it will barf on commas. 

Also, is your postgres running on port 30000?  The standard port is 5432.

No, in order to change template variables you have to be editing the template, which is in another section of the Zabbix UI.  It sounds like you are changing the right thing.

--Rob

**Comment by nawazahmed1** [comment]
Hi Rob,

Yes, I am using the "monitoring" user, but the documentation says that two users can be used if there is a local monitoring agent ? as per [here](http://cavaliercoder.com/libzbxpgsql/documentation/local-monitoring-agent/) in the end of the page This is why i used > user=monitoring,fsep

> pg.connect[user=**monitoring,postgres**]

Yes, I am using the port 30000, and there is no way to change that i guess. ? I ran the below command and got some errors mentioned further below.

zabbix_agentd -p -c /etc/zabbix/zabbix_agentd.conf

zabbix_agentd [6762]: Error: PostgreSQL: could not connect to server: No such file or directory
        Is the server running locally and accepting
        connections on Unix domain socket "/tmp/.s.PGSQL.27500"?


The port 27500 was the initial port setting i had for the postgresql on this server which i later changed to 30000. This server (node2, ip=10.44.37.54) has a zabbix_agent.d installed and a running postgresql cluster (on port 30000). The zabbix server is installed on another server (feplnx7, ip=10.44.37.53) whose ip address is given in the parameters (Server and ServerActive) of server node2's zabbix_agentd.conf. The zabbix_sever host has a postgresql cluster running on port 28000, which also has the "zabdb" database with the schema objects of zabbix monitoring. Additionally, i can see the below errors in the zabbix_agentd.log on node2

 3856:20170224:152635.356 failed to accept an incoming connection: connection from "10.44.37.64" rejected, allowed hosts: "10.44.37.53"
  3856:20170224:152734.803 failed to accept an incoming connection: connection from "10.44.37.64" rejected, allowed hosts: "10.44.37.53"
  3859:20170224:152748.245 no active checks on server [10.44.37.53:10051]: host [node2] not monitored

and these errors from zabbix_server.log of the feplnx7 server.

  4199:20170224:152547.793 cannot send list of active checks to "10.44.37.54": host [node2] not monitored
  4199:20170224:152748.031 cannot send list of active checks to "10.44.37.54": host [node2] not monitored
  4202:20170224:152948.111 cannot send list of active checks to "10.44.37.54": host [node2] not monitored

I can see that the connections from 10.44.37.**64** are getting rejected and only allowed host is "10.44.37.53" in the zabbix_agent.log which is true. I don't even know whose PC/server (10.44.37.64) that is. It must be someone in my corporate network. I can also understand the message from the zabbix_server.log because the status for the node2 on the Configuration -> Hosts (subtab) was disabled but after i enabled it, i see the below messages in zabbix_server.log and there is nothing after that. I had restarted the zabbix_server, httpd and zabbix_agent after I changed the status of node2 to "enabled" on the Configuration -> Hosts (subtab). Is that the reason why the zabbix_server is emitting that last line "node2: host unavailable" when it tried to reach out to node2 for any checks ?

  8408:20170224:153404.623 Zabbix agent item "vm.memory.size[available]" on host "node2" failed: another network error, wait for               15 seconds
  8408:20170224:153422.644 temporarily disabling Zabbix agent checks on host "node2": host unavailable

~
Nawaz

**Comment by nawazahmed1** [comment]
Okay, here is what i further did. I set the parameter PGPORT before invoking the zabbix_agentd -p -c zabbix_agentd.conf and I got the below FATAL errors, which is a good thing. Now i can talk to the postgresql cluster but cannot find the database "fsep" because i think it expects postgres as the database name by default and my configuration is forcing it to look for "fsep" database. In fact, i want it to monitor a database called "db1" as mentioned in the {$PG_DB} => db1 macro variabl.

zabbix_agentd [8262]: Error: PostgreSQL: FATAL:  database "fsep" does not exist (10571)

However, I continue to see the below in zabbix_server.log even after enabling the status to "enabled'.

[root@feplnx7 ~]# tail -f /var/log/zabbix/zabbix_server.log
  8409:20170224:161625.989 cannot send list of active checks to "10.44.37.54": host [node2] not monitored
  8404:20170224:161804.072 Zabbix agent item "system.uptime" on host "node2" failed: first network error, wait for 15 seconds
  8408:20170224:161822.526 Zabbix agent item "vm.memory.size[available]" on host "node2" failed: another network error, wait for 15 seconds
  8408:20170224:161840.543 Zabbix agent item "vm.memory.size[available]" on host "node2" failed: another network error, wait for 15 seconds
  8408:20170224:161858.568 temporarily disabling Zabbix agent checks on host "node2": host unavailable

and the below errors in zabbix_agentd.log

[root@node2 ~]# tail -f /var/log/zabbix/zabbix_agentd.log
  7270:20170224:161644.855 failed to accept an incoming connection: connection from "10.44.37.64" rejected, allowed hosts: "10.44.37.53"
  7270:20170224:161744.869 failed to accept an incoming connection: connection from "10.44.37.64" rejected, allowed hosts: "10.44.37.53"
  7270:20170224:161844.561 failed to accept an incoming connection: connection from "10.44.37.64" rejected, allowed hosts: "10.44.37.53"



I know i am goofing up somewhere but do not exactly know where, will keep trying and keep you updated.

~

Nawaz

**Comment by robbrucks** [comment]
I think you're encountering multiple issues all at the same time, which is making diagnosing it difficult.

Here's what I would do.

1. Login to your Zabbix UI and "unlink and clear" the PostgreSQL template from your host. Make sure the appropriate OS template is linked to your host.  Then you can begin diagnosing any agent/server connectivity issues.  Get the basic monitoring working before you add PostgreSQL monitoring.  Ensure you are not getting any errors in your agent logs and that the Zabbix UI shows your server with a green "ZBX" icon next to it.
2. Then add back the PostgreSQL template and repeat the diagnostic process.

**Comment by nawazahmed1** [comment]
Rob,

I did as per your suggestion, i unlinked and cleared the "Template App PostgreSQL" and deleted the node. I added the node back in but i did not see the "ZBX" icon beside the host as green but it was still grey. I can see the below error in the agent's log.

5983:20170301:171154.108 active check configuration update from [10.44.37.60:10051] started to fail (cannot connect to [[10.44.37.60]:10051]: [111] Connection refused)

However, after i added a "Template OS Linux" for this host in the host's list, the "ZBX" icon turned green after a while. But i still continue to see the error pasted above in the agent log, even though i can see the below entry in the server log.

20170301:171814.977 enabling Zabbix agent checks on host "node2": host became available

Here are the settings of the configuration file, there may be a change in the ip addresses in the servers configured in Zabbix compared to my initial update to this discussion thread.

**zabbix_server.conf** on a host named feplnx7 (ip: 10.44.37.59)
DBHost=10.44.37.59
DBName=zabdb
DBUser=zab
DBPassword=password
DBPort=28000

**zabbix.conf.php** on a host named feplnx7 (ip: 10.44.37.59)

$DB['SERVER']   = '10.44.37.59';
$DB['PORT']     = '28000';
$DB['DATABASE'] = 'zabdb';
$DB['USER']     = 'zab';
$DB['PASSWORD'] = 'password';

$ZBX_SERVER      = '10.44.37.59';
$ZBX_SERVER_PORT = '10051';
$ZBX_SERVER_NAME = 'feplnx7';

zabbix_agentd.conf on a host named node2(ip:10.44.37.60)
Server=10.44.37.59
ServerActive=10.44.37.60
Hostname=node2

Please confirm, if this is correct for the basic monitoring of the host to work neglecting anything to do with PostgreSQL, Also, is this question not fit to be asked on a libzbxpgsql thread and only fit for the zabbix forum ?

Regards,
Nawaz

**Comment by nawazahmed1** [comment]
Hi Rob,

With the above mentioned changes, I let the server run with a bit of a database load just to keep the server busy. I can see the latest data of the server with respect to CPU, Filesystem, General, Memory, OS and Performance etc. In spite of all this, i still see the "(cannot connect to [[10.44.37.60]:10051]: [111] Connection refused)" on the zabbix_agent log. And i see the below entries in zabbix_server.log

7685:20170301:174148.632 executing housekeeper
  7685:20170301:174148.670 housekeeper [deleted 0 hist/trends, 0 items, 0 events, 0 sessions, 0 alarms, 0 audit items in 0.028109 sec, idle for 1 hour(s)]
  7685:20170301:184150.477 executing housekeeper
  7685:20170301:184150.485 housekeeper [deleted 0 hist/trends, 0 items, 0 events, 0 sessions, 0 alarms, 0 audit items in 0.003322 sec, idle for 1 hour(s)]


~

Nawaz

**Comment by robbrucks** [comment]
Based on your post it looks like your agent is misconfigured.

In a typical configuration the "Server" and "ServerActive" parameters on the agent should both point to your Zabbix Server.

So you should probably change the "ServerActive" parameter in the zabbix_agent.conf file on node2 to "10.44.37.59", then restart the agent.

**Comment by nawazahmed1** [comment]
Hi Rob,

Sorry for the late reply. 

I changed the Server and ServerActive directives in zabbix_agent.conf to point to the ip address of the zabbix server. I got the zabbix basic monitoring working and then i imported the libzbxpgsql template and i got that working as well. The moment I changed intervals on the "Discover Rules" from 1 hr to 5 mins. and waited for 5 minutes, I could see the self discovery at work as the number of graphs for node2 rose to over 1700 for all the tables and indexes etc.

I am in the processes of adding the basic/simple monitoring for the tablespaces, I remember reading from the document that the auto-discovery rules will fail for the tablespaces and a separate macro/key item has to be added to get that done. I will give it a try following the documentation or from other discussion threads in the mean time.

Thanks & Best Regards,
Nawaz
